src/Storages/MergeTree/ReplicatedMergeTreeAttachThread.cpp

```diff
@@ -0,0 +1,213 @@
+                    zookeeper->get(zookeeper_path + "/metadata", &metadata_stat);
+                    storage.metadata_version = metadata_stat.version;
+                }
+            });
+
+        /// Temporary directories contain uninitialized results of Merges or Fetches (after forced restart),
+        /// don't allow to reinitialize them, delete each of them immediately.
+        storage.clearOldTemporaryDirectories(0, {"tmp_", "delete_tmp_", "tmp-fetch_"});
+        storage.clearOldWriteAheadLogs();
+        if (storage.getSettings()->merge_tree_enable_clear_old_broken_detached)
+            storage.clearOldBrokenPartsFromDetachedDirecory();
+
+        withRetries([&] { storage.createNewZooKeeperNodes(zookeeper); });
+        withRetries([&] { storage.syncPinnedPartUUIDs(zookeeper); });
+
+        withRetries([&] { storage.createTableSharedID(zookeeper); });
+
+        finalizeInitialization();
+        notifyIfFirstTry();
+    }
+    catch (const Exception & e)
+    {
+        if (e.code() == ErrorCodes::ABORTED && shutdown_called)
+        {
+            LOG_WARNING(log, "Shutdown called, cancelling initialization");
+            return;
+        }
+
+        LOG_ERROR(log, "Initialization failed, table will remain readonly. Error: {}", e.message());
+
+        {
+            std::lock_guard lock(storage.initialization_mutex);
+            storage.init_phase = StorageReplicatedMergeTree::InitializationPhase::INITIALIZATION_DONE;
+        }
+
+        notifyIfFirstTry();
+    }
+};
+
+void ReplicatedMergeTreeAttachThread::finalizeInitialization()
+{
+
+    LOG_INFO(log, "Table is initialized");
+    using enum StorageReplicatedMergeTree::InitializationPhase;
+    {
+        std::lock_guard lock(storage.initialization_mutex);
+        if (!storage.startup_called)
+        {
+            storage.init_phase = INITIALIZATION_DONE;
+            return;
+        }
+
+        storage.init_phase = STARTUP_IN_PROGRESS;
+    }
+
+    storage.startupImpl(std::move(zookeeper));
+}
+
+void ReplicatedMergeTreeAttachThread::notifyIfFirstTry()
+{
+    if (!first_try_done.exchange(true))
+        first_try_done.notify_one();
+}
+
+void ReplicatedMergeTreeAttachThread::tryReconnect()
+{
+    zookeeper = nullptr;
+    while (!shutdown_called)
+    {
+        try
+        {
+            const auto context = storage.getContext();
+            if (storage.zookeeper_name == storage.default_zookeeper_name)
+                zookeeper = std::make_shared<zkutil::ZooKeeper>(context->getConfigRef(), "zookeeper", context->getZooKeeperLog());
+            else
+                zookeeper = context->getAuxiliaryZooKeeperConnection(storage.zookeeper_name);
+            return;
+        }
+        catch (...)
+        {
+            notifyIfFirstTry();
+            LOG_WARNING(log, "Will try to reconnect to ZooKeeper in {} seconds", retry_period);
+            std::this_thread::sleep_for(std::chrono::seconds(retry_period));
```

Comment:
It will occupy one thread in background schedule pool. Better call `task->scheduleAfter(retry_period)` and exit (it will require some changes in `ReplicatedMergeTreeAttachThread::run` to support re-running).

Reply:
yes, the easiest thing would be to just restart everything